contracts/Presale.sol

```
@@ -57,56 +60,84 @@ contract Presale is Ownable {
         vestingDuration = _vestingDuration;
         raiseToken = _raiseToken;
         daoMultisig = _daoMultisig;
     }
 
-    function depositFor(address account, uint256 amount, bytes memory inviteCode, bytes32[] calldata merkleProof) external {
+    /**
+     * @notice Will deposit `amount` of token `raisedToken` for this Presale Round
+     */
+    function depositFor(address account, uint256 amount, uint256 minInvestment, uint256 maxInvestment, bytes memory inviteCode, bytes32[] calldata merkleProof) external {
         require(account != address(0), "Presale: Address cannot be 0x0");
         require(isOpen, "Presale: Round closed");
         require(hardCap > totalAllocated, "Presale: Round closed, goal reached");
-        require(!claimedInvites[inviteCode], "Presale: Invite code has been used");
-        require(MerkleProof.verify(merkleProof, inviteCodesMerkleRoot, keccak256(inviteCode)), "Presale: Invalid invite code");
+
+        /// We only check minInvestment as, if amount excedes maxInvestment it will be scale down.
+        require(allocation[account] >= minInvestment || amount >= minInvestment, "Presale: Can not invest less than minimum amount");
+
+        require(MerkleProof.verify(merkleProof, inviteCodesMerkleRoot, keccak256(abi.encode(inviteCode, minInvestment, maxInvestment))), "Presale: Invalid invite code");
+
+        if (claimedInvites[inviteCode] != address(0)) {
+            require(claimedInvites[inviteCode] == account, "Presale: You can only invest with one wallet per invite code");
+        }
 
         uint256 remainingAllocation = hardCap - totalAllocated;
         if (remainingAllocation < amount) {
             amount = remainingAllocation;
         }
 
+        uint256 remainingInvestment = maxInvestment - allocation[account];
+        if (remainingInvestment < amount) {
+            amount = remainingInvestment;
+        }
+
+        require(amount > 0, "Presale: You have invest up to your limit");
```

Any reason for this extra require? Won't the remaining logic just work with `amount == 0`

As a pattern, require for things that are invariants, any nice to have checks should be handled in the frontend.

---

Okay, leave-in (I'm resolving as such). 

FWIW There are 3 cases and the pattern I'd like to push is require capture invariants (eg. preconditions), with any nice to have useability checks pushed to the DApp. (@ashvds and @samfroggy for visibility/FYI). Moving debate on this to TG.

1. Direct contract interaction via etherscan (super user mode, no need to optimise for usability here)
2. Interaction via the DApp (in which case, the DApp should report the error)
3. Interaction via the DApp, however when a transaction is submitted, the user has an allocation, by the time their transaction is mined, they have lost it.

It's only (3) which is a real usability issue, which (IMHO) should and can be handled via appropriate DApp logic. It's worth investing in building this FE infrastructure up, as once we do, we can re-use it easily. [@timbod7 ](https://github.com/timbod7) has been building some of this up and would be happy to advise I'm sure :)